src/Common/src/System/Net/SafeCloseSocket.Unix.cs

```diff
@@ -20,25 +20,22 @@ internal partial class SafeCloseSocket :
         private int _receiveTimeout = -1;
         private int _sendTimeout = -1;
         private bool _nonBlocking;
+        private SocketAsyncContext _asyncContext;
 
         public SocketAsyncContext AsyncContext
         {
             get
             {
-                return _innerSocket == null ?
-                    SocketAsyncContext.ClosedAsyncContext :
-                    _innerSocket.AsyncContext;
-            }
-        }
+                if (Volatile.Read(ref _asyncContext) == null)
+                {
+                    Interlocked.CompareExchange(ref _asyncContext, new SocketAsyncContext(this, SocketAsyncEngine.Instance), null);
+                }
```

Comment:
I don't remember what state there is on SocketAsyncContext... is there anything to dispose of if we lose the race here?